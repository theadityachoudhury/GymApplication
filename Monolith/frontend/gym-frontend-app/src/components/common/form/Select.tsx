import {
	SelectHTMLAttributes,
	ReactNode,
	useState,
	useRef,
	useEffect,
} from 'react';
import { ChevronDown } from 'lucide-react';

export interface SelectOption {
	value: string;
	label: string;
}

interface SelectProps
	extends Omit<
		SelectHTMLAttributes<HTMLSelectElement>,
		'onChange' | 'size' | 'value'
	> {
	options: SelectOption[];
	label?: string;
	variant?: 'primary';
	size?: 'small' | 'medium' | 'large';
	fullWidth?: boolean;
	isLoading?: boolean;
	leftIcon?: ReactNode;
	rightIcon?: ReactNode;
	className?: string;
	onChange?: (value: string) => void;
	placeholder?: string;
	value?: string;
}

const Select = ({
	options,
	label,
	variant = 'primary',
	size = 'medium',
	fullWidth = false,
	isLoading = false,
	leftIcon,
	rightIcon,
	className = '',
	disabled,
	onChange,
	placeholder = 'Select an option',
	value,
}: SelectProps) => {
	const [isOpen, setIsOpen] = useState(false);
	const [focusedIndex, setFocusedIndex] = useState<number>(-1);
	const selectRef = useRef<HTMLFieldSetElement>(null);

	// Initialize with first option if value is undefined and options exist
	useEffect(() => {
		if (value === undefined && options.length > 0 && onChange) {
			onChange(options[0].value);
		}
	}, [onChange, options, value]);

	// Find the currently selected option based on value prop
	const selectedOption =
		options.find(option => option.value === value) ||
		(options.length > 0 ? options[0] : null);

	// Base styles for the select container
	const baseStyles =
		'rounded-md font-medium transition-all flex items-center justify-between relative';

	// Variant styles
	const variantStyles = {
		primary: 'text-neutral-700 cursor-pointer',
	};

	// Size styles
	const sizeStyles = {
		small: ' py-2 px-3',
		medium: ' py-2 px-4',
		large: ' py-2 px-6',
	};

	// Width styles
	const widthStyles = fullWidth ? 'w-full' : '';

	// Disabled styles
	const disabledStyles =
		disabled || isLoading ? 'opacity-70 cursor-not-allowed' : '';

	// Handle outside clicks to close dropdown
	useEffect(() => {
		const handleClickOutside = (event: MouseEvent) => {
			if (
				selectRef.current &&
				!selectRef.current.contains(event.target as Node)
			) {
				setIsOpen(false);
			}
		};

		document.addEventListener('mousedown', handleClickOutside);
		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, []);

	// Handle option selection
	const handleSelectOption = (option: SelectOption) => {
		setIsOpen(false);
		if (onChange) {
			onChange(option.value);
		}
	};

	// Keyboard navigation
	const handleKeyDown = (e: React.KeyboardEvent) => {
		if (disabled || isLoading) return;

		switch (e.key) {
			case 'ArrowDown':
				e.preventDefault();
				if (!isOpen) {
					setIsOpen(true);
				} else {
					setFocusedIndex(prev =>
						prev < options.length - 1 ? prev + 1 : prev
					);
				}
				break;
			case 'ArrowUp':
				e.preventDefault();
				if (isOpen) {
					setFocusedIndex(prev => (prev > 0 ? prev - 1 : 0));
				}
				break;
			case 'Enter':
			case ' ': // Space
				e.preventDefault();
				if (isOpen && focusedIndex >= 0) {
					handleSelectOption(options[focusedIndex]);
				} else {
					setIsOpen(!isOpen);
				}
				break;
			case 'Escape':
				e.preventDefault();
				setIsOpen(false);
				break;
			default:
				break;
		}
	};

	return (
		<fieldset
			className="flex flex-col rounded-md border text-primary-black px-1 border-neutral-200 focus-within:border-1 focus-within:border-primary-black"
			ref={selectRef}
		>
			{label && (
				<legend className="text-xs text-neutral-700 px-2">
					{label}
				</legend>
			)}
			<div
				className={`
          ${baseStyles}
          ${variantStyles[variant]}
          ${sizeStyles[size]}
          ${widthStyles}
          ${disabledStyles}
          ${className}
        `}
				tabIndex={disabled ? -1 : 0}
				onKeyDown={handleKeyDown}
				onClick={() => !disabled && !isLoading && setIsOpen(!isOpen)}
				role="combobox"
				aria-expanded={isOpen}
				aria-haspopup="listbox"
				aria-disabled={disabled || isLoading}
			>
				{isLoading ? (
					<>
						<svg
							className="animate-spin -ml-1 mr-2 h-4 w-4 text-current"
							xmlns="http://www.w3.org/2000/svg"
							fill="none"
							viewBox="0 0 24 24"
						>
							<circle
								className="opacity-25"
								cx="12"
								cy="12"
								r="10"
								stroke="currentColor"
								strokeWidth="4"
							></circle>
							<path
								className="opacity-75"
								fill="currentColor"
								d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
							></path>
						</svg>
						Loading...
					</>
				) : (
					<>
						{leftIcon && <span className="mr-2">{leftIcon}</span>}
						<span className="flex-grow text-left text-primary-black font-normal truncate">
							{selectedOption
								? selectedOption.label
								: placeholder}
						</span>
						{rightIcon || (
							<ChevronDown
								className={`ml-2 h-3 w-3 transition-transform duration-300 ease-in-out ${isOpen ? 'rotate-180' : ''}`}
							/>
						)}
					</>
				)}

				{/* Dropdown menu */}
				{isOpen && !disabled && !isLoading && (
					<div
						className="absolute left-0 right-0 top-full mt-1 overflow-auto scrollbar-hide z-10 bg-white border border-gray-200 rounded-md shadow-lg"
						role="listbox"
					>
						{options.map((option, index) => (
							<div
								key={option.value}
								className={`
                  px-4 py-2 cursor-pointer hover:bg-green-100 hover:bg-opacity-10
                  ${focusedIndex === index ? 'bg-green-100 bg-opacity-10' : ''}
                  ${selectedOption?.value === option.value ? 'font-medium bg-green-100 bg-opacity-20' : ''}
                `}
								onClick={() => handleSelectOption(option)}
								role="option"
								aria-selected={
									selectedOption?.value === option.value
								}
								onMouseEnter={() => setFocusedIndex(index)}
							>
								{option.label}
							</div>
						))}
					</div>
				)}
			</div>
		</fieldset>
	);
};

export default Select;
