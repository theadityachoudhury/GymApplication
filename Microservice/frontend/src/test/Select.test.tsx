/* eslint-disable @typescript-eslint/no-unused-vars */
import React from 'react';
import {
	render,
	screen,
	fireEvent,
	waitFor,
	within,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Select from '../components/common/form/Select';
import { describe, expect, it } from 'vitest';

// Mock options for testing
const mockOptions = [
	{ value: 'option1', label: 'Option 1' },
	{ value: 'option2', label: 'Option 2' },
	{ value: 'option3', label: 'Option 3' },
];

describe('Select Component', () => {
	// Basic rendering tests
	it('renders with required props', () => {
		render(<Select options={mockOptions} />);

		// Should show first option by default
		expect(screen.getByText('Option 1')).toBeInTheDocument();
	});

	it('renders with custom label', () => {
		render(<Select options={mockOptions} label="Test Label" />);

		expect(screen.getByText('Test Label')).toBeInTheDocument();
	});

	it('renders with placeholder when no value is selected', () => {
		// Prevent auto-selection by providing empty options
		render(<Select options={[]} placeholder="Select something" />);

		expect(screen.getByText('Select something')).toBeInTheDocument();
	});

	// Dropdown functionality tests
	it('opens dropdown when clicked', async () => {
		render(<Select options={mockOptions} />);

		const combobox = screen.getByRole('combobox');
		await userEvent.click(combobox);

		// Check if the listbox is visible
		expect(screen.getByRole('listbox')).toBeInTheDocument();

		// Check for options within the listbox
		const listbox = screen.getByRole('listbox');
		const options = within(listbox).getAllByRole('option');
		expect(options).toHaveLength(3);
		expect(options[0]).toHaveTextContent('Option 1');
		expect(options[1]).toHaveTextContent('Option 2');
		expect(options[2]).toHaveTextContent('Option 3');
	});

	it('closes dropdown when an option is selected', async () => {
		render(<Select options={mockOptions} />);

		// Open dropdown
		await userEvent.click(screen.getByRole('combobox'));

		// Select an option (the second one)
		const listbox = screen.getByRole('listbox');
		const options = within(listbox).getAllByRole('option');
		await userEvent.click(options[1]);

		// Dropdown should be closed
		expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
	});

	it('calls onChange when an option is selected', async () => {
		const handleChange = vi.fn();
		render(<Select options={mockOptions} onChange={handleChange} />);

		// Open dropdown
		await userEvent.click(screen.getByRole('combobox'));

		// Select an option (the second one)
		const listbox = screen.getByRole('listbox');
		const options = within(listbox).getAllByRole('option');
		await userEvent.click(options[1]);

		// onChange should be called with the selected value
		expect(handleChange).toHaveBeenCalledWith('option2');
	});

	// Value and selection tests
	it('displays the selected option based on value prop', () => {
		render(<Select options={mockOptions} value="option2" />);

		const combobox = screen.getByRole('combobox');
		const selectedText = within(combobox).getByText('Option 2');
		expect(selectedText).toBeInTheDocument();
	});

	it('selects first option by default when value is undefined', () => {
		const handleChange = vi.fn();
		render(<Select options={mockOptions} onChange={handleChange} />);

		// onChange should be called with the first option's value
		expect(handleChange).toHaveBeenCalledWith('option1');
	});

	// Disabled state tests
	it('does not open dropdown when disabled', async () => {
		render(<Select options={mockOptions} disabled />);

		await userEvent.click(screen.getByRole('combobox'));

		// Dropdown should not be visible
		expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
	});

	it('applies disabled styling', () => {
		render(<Select options={mockOptions} disabled />);

		expect(screen.getByRole('combobox')).toHaveAttribute(
			'aria-disabled',
			'true'
		);
		expect(screen.getByRole('combobox')).toHaveClass('opacity-70');
	});

	// Loading state tests
	it('shows loading state when isLoading is true', () => {
		render(<Select options={mockOptions} isLoading />);

		expect(screen.getByText('Loading...')).toBeInTheDocument();
		expect(screen.getByRole('combobox')).toHaveAttribute(
			'aria-disabled',
			'true'
		);
	});

	it('does not open dropdown when loading', async () => {
		render(<Select options={mockOptions} isLoading />);

		await userEvent.click(screen.getByRole('combobox'));

		// Dropdown should not be visible
		expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
	});

	// Icon tests
	it('renders with left icon', () => {
		const leftIcon = <span data-testid="left-icon">←</span>;
		render(<Select options={mockOptions} leftIcon={leftIcon} />);

		expect(screen.getByTestId('left-icon')).toBeInTheDocument();
	});

	it('renders with right icon', () => {
		const rightIcon = <span data-testid="right-icon">→</span>;
		render(<Select options={mockOptions} rightIcon={rightIcon} />);

		expect(screen.getByTestId('right-icon')).toBeInTheDocument();
	});

	it('renders default chevron icon when no right icon is provided', () => {
		render(<Select options={mockOptions} />);

		// Check for SVG icon (ChevronDown)
		const svg = document.querySelector('svg');
		expect(svg).toBeInTheDocument();
	});

	// Size tests
	it('applies small size styles', () => {
		render(<Select options={mockOptions} size="small" />);

		expect(screen.getByRole('combobox')).toHaveClass('text-sm');
	});

	it('applies medium size styles', () => {
		render(<Select options={mockOptions} size="medium" />);

		expect(screen.getByRole('combobox')).toHaveClass('text-base');
	});

	it('applies large size styles', () => {
		render(<Select options={mockOptions} size="large" />);

		expect(screen.getByRole('combobox')).toHaveClass('text-lg');
	});

	// Width tests
	it('applies full width when fullWidth is true', () => {
		render(<Select options={mockOptions} fullWidth />);

		expect(screen.getByRole('combobox')).toHaveClass('w-full');
	});

	// Custom class tests
	it('applies custom className', () => {
		render(<Select options={mockOptions} className="custom-class" />);

		expect(screen.getByRole('combobox')).toHaveClass('custom-class');
	});

	// Keyboard navigation tests
	it('opens dropdown with Space key', async () => {
		render(<Select options={mockOptions} />);

		const combobox = screen.getByRole('combobox');
		combobox.focus();
		await userEvent.keyboard(' ');

		expect(screen.getByRole('listbox')).toBeInTheDocument();
	});

	it('opens dropdown with Enter key', async () => {
		render(<Select options={mockOptions} />);

		const combobox = screen.getByRole('combobox');
		combobox.focus();
		await userEvent.keyboard('{Enter}');

		expect(screen.getByRole('listbox')).toBeInTheDocument();
	});

	it('closes dropdown with Escape key', async () => {
		render(<Select options={mockOptions} />);

		// Open dropdown
		const combobox = screen.getByRole('combobox');
		await userEvent.click(combobox);

		// Press Escape
		await userEvent.keyboard('{Escape}');

		// Dropdown should be closed
		expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
	});

	it('navigates options with arrow keys', async () => {
		render(<Select options={mockOptions} />);

		// Open dropdown
		const combobox = screen.getByRole('combobox');
		await userEvent.click(combobox);

		// Press down arrow to focus first option
		await userEvent.keyboard('{ArrowDown}');

		// Get all options
		const listbox = screen.getByRole('listbox');
		const options = within(listbox).getAllByRole('option');

		// First option should have focus styling
		expect(options[0]).toHaveClass('bg-green-100');

		// Press down arrow again to focus second option
		await userEvent.keyboard('{ArrowDown}');

		// Second option should now have focus styling
		expect(options[1]).toHaveClass('bg-green-100');
	});

	it('selects focused option with Enter key', async () => {
		const handleChange = vi.fn();
		render(<Select options={mockOptions} onChange={handleChange} />);

		// Open dropdown
		const combobox = screen.getByRole('combobox');
		await userEvent.click(combobox);

		// Navigate to second option
		await userEvent.keyboard('{ArrowDown}');
		await userEvent.keyboard('{ArrowDown}');

		// Select with Enter
		await userEvent.keyboard('{Enter}');

		// onChange should be called with the second option's value
		expect(handleChange).toHaveBeenCalledWith('option2');

		// Dropdown should be closed
		expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
	});

	// Outside click tests
	it('closes dropdown when clicking outside', async () => {
		render(
			<div>
				<Select options={mockOptions} />
				<div data-testid="outside">Outside</div>
			</div>
		);

		// Open dropdown
		await userEvent.click(screen.getByRole('combobox'));

		// Dropdown should be visible
		expect(screen.getByRole('listbox')).toBeInTheDocument();

		// Click outside
		await userEvent.click(screen.getByTestId('outside'));

		// Dropdown should be closed
		expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
	});

	// Edge cases
	it('handles empty options array', () => {
		render(<Select options={[]} />);

		// Should show placeholder
		expect(screen.getByText('Select an option')).toBeInTheDocument();
	});

	it('highlights selected option in dropdown', async () => {
		render(<Select options={mockOptions} value="option2" />);

		// Open dropdown
		await userEvent.click(screen.getByRole('combobox'));

		// Get all options
		const listbox = screen.getByRole('listbox');
		const options = within(listbox).getAllByRole('option');

		// Second option should have selected styling
		expect(options[1]).toHaveClass('font-medium');
		expect(options[1]).toHaveClass('bg-green-100');
	});

	// Accessibility tests
	it('has correct ARIA attributes', () => {
		render(<Select options={mockOptions} />);

		const combobox = screen.getByRole('combobox');
		expect(combobox).toHaveAttribute('aria-expanded', 'false');
		expect(combobox).toHaveAttribute('aria-haspopup', 'listbox');

		// Open dropdown
		fireEvent.click(combobox);

		expect(combobox).toHaveAttribute('aria-expanded', 'true');
		expect(screen.getByRole('listbox')).toBeInTheDocument();

		// Options should have aria-selected
		const listbox = screen.getByRole('listbox');
		const options = within(listbox).getAllByRole('option');
		expect(options[0]).toHaveAttribute('aria-selected', 'true'); // First option is selected by default
		expect(options[1]).toHaveAttribute('aria-selected', 'false');
	});

	it('is keyboard focusable', async () => {
		render(<Select options={mockOptions} />);

		// Tab to focus the select
		await userEvent.tab();

		expect(screen.getByRole('combobox')).toHaveFocus();
	});

	// Snapshot test
	it('matches snapshot', () => {
		const { container } = render(
			<Select options={mockOptions} label="Test Select" />
		);
		expect(container).toMatchSnapshot();
	});
});
