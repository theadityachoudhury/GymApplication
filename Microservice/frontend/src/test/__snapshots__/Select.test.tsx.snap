// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select Component > matches snapshot 1`] = `
<div>
  <fieldset
    class="flex flex-col rounded-md border text-primary-black px-1 border-neutral-200 focus-within:border-1 focus-within:border-primary-black"
  >
    <legend
      class="text-xs text-neutral-700 px-2"
    >
      Test Select
    </legend>
    <div
      aria-disabled="false"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="
          rounded-md font-medium transition-all flex items-center justify-between relative
          text-neutral-700 cursor-pointer
          text-base py-3 px-4
          
          
          
        "
      role="combobox"
      tabindex="0"
    >
      <span
        class="flex-grow text-left text-primary-black font-normal truncate"
      >
        Option 1
      </span>
      <svg
        class="lucide lucide-chevron-down ml-2 h-5 w-5 transition-transform duration-300 ease-in-out"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </div>
  </fieldset>
</div>
`;
